### PR TITLE
Refactor PiranhaJava configuration internals.

### PR DESCRIPTION
**Pull request by lazaroclapp**
This change:
a) Switches `Map<String, List<Map<String, Object>>>` for an immutable map
   of structured records (PiranhaMethodRecord) [#54]
b) Performs ALL configuration parsing at init() time.
c) Removes confused mixed uses of two ParseException classes (one from
  `java.text.` and the other from `org.json.simple.parser.` and
   introduces our own `PiranhaConfigurationException` to encapsulate
   everything that can go wrong during Piranha configuration/initialization.
d) Adds a WARNING when seeing an old-format `piranha.properties` file. It
   adds a test for this. [#49]
e) Removes `throws` clause from the `XPFlagCleaner` constructor, as that
   exception can no longer happen there (it must happen during the latter
   call to `init()` and Piranha just silently disables itself in that case)

The change to `XPFlagCleanerTest` is huge, but mostly just removing redundant
try{...}catch{...} blocks for the never thrown exception in (e).

### CODE REVIEW

File: java/piranha/src/test/java/com/uber/piranha/XPFlagCleanerTest.java [hunk 1]
```diff
@@ -2237,16 +2044,34 @@ public void exceptionWhenConfigFileInvalid() throws Exception {
...
+
+    XPFlagCleaner flagCleaner = new XPFlagCleaner();
+    flagCleaner.init(b.build());
+  }
+
+  /*
+   * Check that the user is warned when using an old style piranha.properties file
+   * */
+  @Test
+  public void warningWhenConfigFileOld() throws Exception {
```
Comment (lazaroclapp) [inline java/piranha/src/test/java/com/uber/piranha/XPFlagCleanerTest.java:2058]: This test is new, everything else in this file is removing old `try{...}catch{...}` blocks and updating error messages.

File: java/piranha/src/test/resources/config/invalid/piranha.properties [hunk 2]
```diff
@@ -0,0 +1,6 @@
+treatedMethods=isToggleEnabled,isFlagTreated
+controlMethods=isToggleDisabled
+emptyMethods=putToggleEnabled,putToggleDisabled,includeEvent
+treatmentGroupMethods=isToggleInGroup
+annotations=ToggleTesting
+linkURL=<provide_your_url>
```
Comment (lazaroclapp) [inline java/piranha/src/test/resources/config/invalid/piranha.properties:6]: Will fix this missing new line.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 3]
```diff
@@ -349,23 +335,20 @@ private API getXPAPI(ExpressionTree et) {
...
-      MethodInvocationTree mit, List<Map<String, Object>> methodPropertiesForMethodName) {
-    for (Map<String, Object> currentMethodProperties : methodPropertiesForMethodName) {
-      Integer argumentIndex = getArgumentIndexFromMap(currentMethodProperties);
-
+      MethodInvocationTree mit, ImmutableCollection<PiranhaMethodRecord> methodRecordsForName) {
+    for (PiranhaMethodRecord methodRecord : methodRecordsForName) {
       // when argumentIndex is specified, if mit's argument at argIndex doesn't match xpFlagName,
       // skip to next method property map
-      if (argumentIndex != null) {
+      if (methodRecord.getArgumentIdx().isPresent()) {
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:350]: Can we store the `methodRecord.getArgumentIdx()` locally and re-use it to avoid verbosity?

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 4]
```diff
@@ -298,7 +305,7 @@ public Description matchCompilationUnit(
     if (!initialized && !disabled) {
       try {
         init(flags);
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:307]: just noticing this even though not part of this change. Why was this under `matchCompilationUnit`? Ideally, `init` at the beginning right after construction would be better. Is this to handle the silent exception problem in the constructor?

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 5]
```diff
@@ -379,37 +362,26 @@ private API getXPAPI(
       MemberSelectTree mst = ((MemberSelectTree) mit.getMethodSelect());
       // when returnType is specified, check if mst's return type matches it
       // if it's not a match, skip to next method property map
-      String returnType = getValueStringFromMap(currentMethodProperties, RETURN_TYPE_STRING);
-      if (returnType != null) {
+      if (methodRecord.getReturnType().isPresent()) {
         String mReturn = ASTHelpers.getReturnType(mst).toString();
-        if (!returnType.equals(mReturn)) {
+        if (!methodRecord.getReturnType().get().equals(mReturn)) {
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:367]: Store `methodRecord.getReturnType()` to avoid verbosity?

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java [hunk 6]
```diff
@@ -379,37 +362,26 @@ private API getXPAPI(
...
         }
       }
       // when receiverType is specified, check if mst's receiver type matches it
       // if it's not a match, skip to next method property map
-      String receiverType = getValueStringFromMap(currentMethodProperties, RECEIVER_TYPE_STRING);
-      if (receiverType != null) {
+      if (methodRecord.getReceiverType().isPresent()) {
         String mReceive = ASTHelpers.getReceiverType(mst).toString();
-        if (!receiverType.equals(mReceive)) {
+        if (!methodRecord.getReceiverType().get().equals(mReceive)) {
```
Comment (mkr-plse) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:375]: verbosity:?

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java (hunk 3 above)
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:350]: I don't think it saves too much. We'd be replacing a call to `.getArgumentIdx()` with adding a new `Optional<Integer>` local. Either way we need to check for `isPresent()` and then do `get()` to get the actual value, so this seemed clearer to me. That said, I don't feel strongly either way, can rewrite to introduce the local.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java (hunk 5 above)
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:367]: See above.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java (hunk 6 above)
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:375]: See above.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java (hunk 4 above)
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:307]: This is exactly to avoid the silent swallowing of the error inside the constructor and use of the zero-arg constructor. I think it is safer and easier to debug this way, but it is a bit of a workaround for the EP's "feature" of silent fallback of constructors :/ 

It's also because of this that I couldn't make the `ImmutableMultimap` also `final`. But I do think it's the best we can do right now.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java (hunk 3 above)
Comment (mkr-plse) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:350]: I am not referring to the performance (the compiler will likely take care of this) but to the readability. Feel free to update (or not).

Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:350]: I am also not talking about performance. I am just not sure it's more readable/less verbose/less idiomatic to go `var.getField.isPresent()` and then `T field = var.getField.get()` vs `Optional<T> maybeField = var.getField()` then `maybeField.isPresent` then `T field = maybeField.get()`.

But I have no strong opinion here, so let me update.
